frame/democracy/src/lib.rs

```diff
@@ -588,22 +604,34 @@ decl_module! {
 		///
 		/// - `proposal_hash`: The hash of the proposal preimage.
 		/// - `value`: The amount of deposit (must be at least `MinimumDeposit`).
+		/// - `prop_count`: Witness data which must be at least `PublicProps::<T>::decode_len()`
 		///
 		/// Emits `Proposed`.
 		///
-		/// # <weight>
-		/// - Complexity: `O(1)`
-		/// - Db reads: `PublicPropCount`, `PublicProps`
-		/// - Db writes: `PublicPropCount`, `PublicProps`, `DepositOf`
-		/// # </weight>
-		#[weight = T::WeightInfo::propose()]
-		fn propose(origin, proposal_hash: T::Hash, #[compact] value: BalanceOf<T>) {
+		/// Weight: `O(p)`
+		#[weight = T::WeightInfo::propose(*prop_count)]
+		fn propose(origin,
+			proposal_hash: T::Hash,
+			#[compact] value: BalanceOf<T>,
+			#[compact] prop_count: u32,
```

Comment:
Taking a second look at this function, I think it is fair to call this O(1).

We use `decode_len` in order to figure out the length, which should not need us to decode the whole vector, just the first bytes, then we use `append` to add the new public proposal, which again does not decode/encode the existing proposals, just places it on the end already encoded.

If we look at the benchmark results itself, it is saying that each proposal adds 11 nanoseconds, which is extremely small and within the errors we would expect for the benchmarking pipeline.

I think it would be perfectly safe to keep treating this as O(1) and remove the `prop_count`.

What do you think?